improver/ensemble_copula_coupling/utilities.py

```diff
@@ -289,3 +259,100 @@ def restore_non_percentile_dimensions(
     if n_percentiles > 1:
         shape_to_reshape_to = [n_percentiles] + shape_to_reshape_to
     return array_to_reshape.reshape(shape_to_reshape_to)
+
+
+def flatten_and_interpolate(
+    desired_percentiles: Union[list, ndarray],
+    original_percentiles: ndarray,
+    data: ndarray,
+    max_value: float = 100,
+) -> ndarray:
+    """
+    Interpolates percentiles or probabilities in a semi-vectorised manner to
+    provide greater speed. The interpolation is performed on a cube with
+    3-dimensions. The last dimension is equivalent to xp in usual interpolation
+    syntax. The leading dimension is looped over as a compromise between
+    vectorisation and memory requirements given typical array sizes. Between
+    these the dimensions are collapsed and the interpolation operation is
+    vectorised.
+
+    The method works by applying the numpy interpolation functionality to large
+    1-dimensional arrays. These are constructed by repeating the original and
+    desired percentile sequences N times in 1-dimensional arrays, where N is the
+    length of the grid dimension being vectorised. The data taken from different
+    grid points along the dimension being vectorised that corresponds to the
+    original percentiles is also placed end-to-end in a single 1-dimensional
+    array. Finally, each repeated sequence of percentiles in both the original
+    and desired sets is offset by some value (max_value). Interpolation is then
+    used with the three long 1-dimensional arrays in a single step.
+
+    For a typical spatial grid the vectorisation applied along say x, but a loop
+    is used over y. This reduces the amount of memory used whilst providing a
+    speed up.
+
+    This method will become slower for increasingly large percentile coordinates
+    but provides a significant speed up for typical numbers of percentiles
+    (i.e. up to around 20).
+
+    Args:
+        desired_percentiles:
+            The target percentiles.
+        original_percentiles:
+            The percentiles at which the data is provided.
+        data:
+            The data to be interpolated.
+        max_value:
+            The maximum potential value of the percentile coordinate, defaults
+            to 100. Used to calculate an offset for vectorisation.
+
+    Returns:
+        The forecast values interpolated to the desired_percentiles.
+    """
+
+    def calculate_offset(percentiles_size: int, dim_size: int):
+        """Calculate offsets, each step has a length determined by the variable
+        percentile_size. The number of steps is set by dim_size. The increment
+        is set by max_value. An array describing the offsets is returned."""
+        steps = []
+        for i in range(0, dim_size):
+            steps.extend(np.ones(percentiles_size, dtype=np.int) * i * max_value)
+        return np.array(steps).reshape(percentiles_size, dim_size)
```

Comment:
I think it would be more efficient to pre-allocate ``steps``:
```
        steps = np.ones((percentiles_size, dim_size), dtype=np.int) * max_value
        for i in range(0, dim_size):
            steps[:, i] *= i
        return steps
```

Reply:
Done.